rootston/desktop.c

```diff
@@ -755,6 +758,52 @@ static void input_inhibit_deactivate(struct wl_listener *listener, void *data) {
 	}
 }
 
+static void handle_constraint_inactive(
+		struct wl_listener *listener,
+		struct wlr_pointer_constraint_v1 *constraint) {
+	struct roots_seat *seat = constraint->seat->data;
+	struct roots_cursor *cursor = seat->cursor;
+
+	if (constraint->type == WLR_POINTER_CONSTRAINT_V1_CONFINED) {
+		wlr_cursor_confine_to_region(cursor->cursor, WLR_BOX_EMPTY);
+	} else {
+		struct roots_view *view = constraint->surface->data;
+
+		seat->cursor->locked = false;
+		if (constraint->current.cursor_hint.valid) {
+			wlr_cursor_warp(cursor->cursor, NULL,
+				view->x + constraint->current.cursor_hint.x,
+				view->y + constraint->current.cursor_hint.y);
+		}
+	}
+}
+
+static void handle_constraint_active(
+		struct wl_listener *listener,
+		struct wlr_pointer_constraint_v1_activation_event *activation) {
+	struct wlr_box box = activation->box;
+	struct wlr_pointer_constraint_v1 *constraint = activation->pointer_constraint;
+	struct roots_seat* seat = constraint->seat->data;
+
+	if (constraint->type == WLR_POINTER_CONSTRAINT_V1_CONFINED) {
+		struct roots_view *view = constraint->surface->data;
+		if (wlr_box_empty(&box)) {
+			view_get_box(view, &box);
+		} else {
+			box.x += view->x;
+			box.y += view->y;
+		}
+		wlr_cursor_confine_to_region(seat->cursor->cursor, box);
```

Comment:
What if the view moves?

Reply:
(This is one of the reasons why I'd prefer the constraint stuff to be out of `wlr_cursor` and managed by rootston)

Reply:
emersion, I have come to the conclusion that it's very impractical to have this be in rootston. Handling locked pointers is trivial in rootston, but confines are quite complex, if you want to handle anything other than the basic single rectangle region case. If the view moves, the compositor must simply call it again. I have nearly 100 lines of code for handling cursor moves, and this has to be done for other types of cursor movement too, like absolute motion. Counting on users of this library to properly implement this is not optimal at all IMO.